pkg/storage/fingerprint_writer.go

```diff
@@ -0,0 +1,260 @@
+import (
+	"context"
+	"hash"
+	"io"
+
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/util/uuid"
+	"github.com/cockroachdb/errors"
+)
+
+// fingerprintWriter hashes every key/timestamp and value for point keys, and
+// combines their hashes via a XOR into a running aggregate.
+//
+// Range keys are not fingerprinted but instead written to a pebble SST that is
+// returned to the caller. This is because range keys do not have a stable,
+// discrete identity and so it is up to the caller to define a deterministic
+// fingerprinting scheme across all returned range keys.
+//
+// The caller must Finish() and Close() the fingerprintWriter to finalize the
+// writes to the underlying pebble SST.
+type fingerprintWriter struct {
+	hasher hash.Hash64
+
+	sstWriter *SSTWriter
+	xorAgg    *uintXorAggregate
+}
+
+// makeFingerprintWriter creates a new fingerprintWriter.
+func makeFingerprintWriter(
+	ctx context.Context, hasher hash.Hash64, cs *cluster.Settings, f io.Writer,
+) fingerprintWriter {
+	// TODO(adityamaru,dt): Once
+	// https://github.com/cockroachdb/cockroach/issues/90450 has been addressed we
+	// should write to a kvBuf instead of a Backup SST writer.
+	sstWriter := MakeBackupSSTWriter(ctx, cs, f)
+	return fingerprintWriter{
+		sstWriter: &sstWriter,
+		hasher:    hasher,
+		xorAgg:    &uintXorAggregate{},
+	}
+}
+
+type uintXorAggregate struct {
+	sum uint64
+}
+
+// add inserts one value into the running xor.
+func (a *uintXorAggregate) add(x uint64) {
+	a.sum = a.sum ^ x
+}
+
+// result returns the xor.
+func (a *uintXorAggregate) result() uint64 {
+	return a.sum
+}
+
+// Finish finalizes the underlying SSTWriter, and returns the aggregated
+// fingerprint for point keys.
+func (f *fingerprintWriter) Finish() (uint64, error) {
+	if err := f.sstWriter.Finish(); err != nil {
+		return 0, err
+	}
+	return f.xorAgg.result(), nil
+}
+
+// Close finishes and frees memory and other resources. Close is idempotent.
+func (f *fingerprintWriter) Close() {
+	if f.sstWriter == nil {
+		return
+	}
+	f.sstWriter.Close()
+	f.hasher.Reset()
+	f.xorAgg = nil
+	f.sstWriter = nil
+}
+
+var _ Writer = &fingerprintWriter{}
+
+// PutRawMVCCRangeKey implements the Writer interface.
+func (f *fingerprintWriter) PutRawMVCCRangeKey(key MVCCRangeKey, bytes []byte) error {
+	// We do not fingerprint range keys, instead, we write them to a Pebble SST.
+	// This is because range keys do not have a stable, discrete identity and so
+	// it is up to the caller to define a deterministic fingerprinting scheme
+	// across all returned range keys.ler to decide how to fingerprint them.
+	return f.sstWriter.PutRawMVCCRangeKey(key, bytes)
+}
+
+// PutRawMVCC implements the Writer interface.
+func (f *fingerprintWriter) PutRawMVCC(key MVCCKey, value []byte) error {
+	defer f.hasher.Reset()
+
+	// Hash the key/timestamp and value of the RawMVCC.
+	_, err := f.hasher.Write(key.Key)
```

Comment:
This can happen in a follow-up, but I think for the purpose of tenant fingerprinting, we'll eventually want a way to specify whether the tenant prefix of the key and the checksum of the value are included in the fingerprint.

Reply:
Oh good point, I hadn't thought about this. Let me file an issue for this so I don't forget.